Preprocessor for ULP/RTC macros

wnienhaus opened the pull request:
This PR implements issue #19 and in a way also issue #12, because now many examples from the net can be assembled unmodified.

This is a big PR. I therefore split it up into many small commits, each standing on its own and intended to be short and simple to review. It will be best to review this PR one commit at a time. Each commit describes in more detail what the change is.

Each commit has at least unit tests (that also pass) covering the change. All changes also have compat tests (even though they sometimes have their own commit), to ensure the resulting binary always matches what binutils-esp32ulp would have produced. There are also a new set of compat tests, that download sources from the net (including the examples from the tests of binutils-esp32ulp) and run them through both py-esp32-ulp and binutils-esp32ulp to ensure the output is the same).

In summary, the PR has 4 changes:
* Fixes encountered during testing or to support the change.
* Adding support for expression evaluation to the assembler (e.g. 1+1 or 1<<symbol)
* Adding a preprocessor for preprocessing ESP-IDF macros/constants before the code goes to the assembler
* Add a new suite of compat tests using example code from the net as input

I originally set myself the goal to be able to compile all examples from this repo (https://github.com/duff2013/ulptool/tree/master/src/ulp_examples) without modification to the sources. I will say I managed (including it working on the ESP32 itself), except for the cases that use assembler macros (see issue #40), which I intentionally did not implement. (I might try that as a future side project).

The back story to this all is that I used to do a lot of manual steps with my ULP assembly source code. I passed it through the gcc preprocessor to get the macros I used expanded, but those macros spit out long expressions with countless brackets. Manually evaluating those (especially when using my own constants in the expressions) was tedious. I also kept the original code (before preprocessing) in comments above the resulting statements, so I could more easily understand where the final reg_rd or reg_wr statements came from. While that helped, keeping the statements and comments in sync was also messy.

Ultimately, I wanted a better way. I never used binutils-esp32ulp to compile my ULP sources, because when I started with ULP coding, I didn't want to download and build this massive tool on my machine. I found py-esp32-ulp and it was just so simple and fast and did what I needed. So while using biunutils-esp32ulp is an alternative (especially, when creating a docker image of it), I wanted to continue with the lightweight and simple approach of py-esp32-ulp, directly in MicroPython. (Of course this also seemed like a fun side project, which is definitely was! In particular, getting it to work on the actual device with little memory!)

wnienhaus commented:
Actually... it fixes issue #16 too (commit with subject: "add support for the .global directive. only symbols flagged as global will be exported")

ThomasWaldmann commented:
Wow, this is big. :-)

Guess I'll need a bit for reviewing this. Could you please work on #44 while I do that?

Would be good to have master branch testing successfully "as is" and then rebase this PR onto new master.

wnienhaus commented:
Will do. Started looking into making the Travis build pass - the failure is because of the old gcc on Ubuntu Trusty. But if you want to move to Github Actions, I will look into that!

ThomasWaldmann commented:
ok, this now needs a rebase onto current master.

ThomasWaldmann commented on .travis.yml:
```diff
@@ -1,11 +1,16 @@
       - pkg-config
+      - bison
+      - flex
+      - build-essentials
```
was b-e missing in your other PR?

wnienhaus replied:
Yes and no.
The image from Github Actions already includes all that is necessary from build-essentials. So no need to apt-get install it.
Here I added it "just in case" because I tested locally with a bionic docker image (not the one from Travis) and found I also needed build-essentials... and since I didnt have a way to test Travis builds without pushing a change here, I just added it "just-in-case". Then I discovered the Travis build didnt run because of not enough credits. :) 
Anyway... with Github Actions it all works.

wnienhaus commented:
I have now rebased this onto master in my fork. As soon as the "dummy PR" (#46) is merged, I will push the rebased code.

ThomasWaldmann commented:
just merged another PR of yours into master, so refresh master and rebase again.

wnienhaus commented:
So. It's rebased now. And all tests pass :) .

ThomasWaldmann commented on tests/assemble.py:
```diff
@@ -16,12 +17,13 @@ def test_parse_line():
+    assert a.parse_line(lines[3]) == (None, 'ld', ('r0', 'r1', '0', ))
+    assert a.parse_line(lines[4]) == (None, 'st', ('r0', 'r1', '0', ))
+    assert a.parse_line(lines[5]) == (None, 'halt', ())
+    assert a.parse_line(lines[6]) == ('end', None, ())
```
looks like the testing code (as before your changes) is not diff-friendly, due to these indexes.

maybe using an iterator or just parsing the whole thing would have been simpler.

wnienhaus replied:
Yip. I intended to keep the code as close to what it was before - which I tend to do to make diff's clean, but of course here it didn't work out. I'm happy to improve this one - still wont be a clean diff, but then for the future it will be easier to make clean changes that.

ThomasWaldmann commented on esp32_ulp/assemble.py:
```diff
@@ -12,9 +12,10 @@
 
 class SymbolTable:
-    def __init__(self, symbols, bases):
+    def __init__(self, symbols, bases, globls):
```
as long as the method is rather simple/short and does not need to call `globals()` python predefined function, you could also name it `globals`.

wnienhaus replied:
I didn't want to reuse the python keyword, and since some assemblers apparently used `.globl` as the directive name (https://sourceware.org/binutils/docs/as/Global.html#Global), I chose that spelling.
But now that you point this out, I realise that it's really only in the `__init__` method where the "globals" word would be used, everywhere else it would be `_globals` anyway, which is very obviously not python's built-in function.
I will change it.

ThomasWaldmann commented on tests/assemble.py:
```diff
@@ -90,8 +107,20 @@ def test_assemble_bss_with_value():
+    assert a.symbols.has_sym('entry')
+
+    exported_symbols = a.symbols.export()
+    assert 'internal' not in [x[1] for x in exported_symbols]
```
```
assert 'internal' not in [symbol for value, symbol in exported_symbols]
```

wnienhaus replied:
nice improvement!

ThomasWaldmann commented on tests/assemble.py:
```diff
@@ -90,8 +107,20 @@ def test_assemble_bss_with_value():
+
+    exported_symbols = a.symbols.export()
+    assert 'internal' not in [x[1] for x in exported_symbols]
+    assert exported_symbols == [(0, 'counter'), (2, 'entry')]
```
well, if you compare the complete thing anyway, the previous assert is not really necessary.
could be replaced by a comment.

wnienhaus replied:
I like the previous line, because I feel it makes the test more obvious in terms of what we're testing for (i.e. it has documentation value). And I prefer encoding documentation into executable code over comments (where possible), because there's no way to check whether the comments actually match what the code does.

But, it's also true that these 2 lines test for the same thing. So I would almost opt for dropping the 2nd line (the exact comparison), because that type of check should already be covered by other tests anyway. What do you think?

wnienhaus replied:
Now looking at the test's code in full, I see in that I am basically testing the one vs the other, so it's very clear what is being tested even without comment. So I will drop the `assert internal in/not in...` lines.

ThomasWaldmann reviewed (commented):
started reviewing:

add units test for the .set directive 5a9b7e9  
add support for left aligned assembler directives (e.g. .set)  … 8d7981d  
fix a crash bug where BSS size calculation was attemped on the value …  … 45b3769  
raise error when attempting to store values in .bss section  … f65cbc2  
fix reference to non-exiting variable f0d86db  
fixed typo in comment of instruction definition 15633eb  
add support for the .global directive. only symbols flagged as global…  … 3a2d312  
let SymbolTable.export() optionally export non-global symbols too  … 2417778

i only found a few nitpicks. overall, great stuff so far. i like how you do clean commits!

wnienhaus commented on tests/assemble.py:
```diff
@@ -53,6 +60,17 @@ def test_assemble():
+    else:
+        raised = False
+    assert not raised
+
```
will do

wnienhaus commented on esp32_ulp/assemble.py:
```diff
@@ -32,7 +32,7 @@ def get_from(self):
         entry = (stype, section, value)
         if symbol in self._symbols and entry != self._symbols[symbol]:
-            raise Exception('redefining symbol %s with different value %r -> %r.' % (label, self._symbols[symbol], entry))
+            raise Exception('redefining symbol %s with different value %r -> %r.' % (symbol, self._symbols[symbol], entry))
```
Good spotting. I do know `git rebase -i` - the main reason why the commits are so clean... they have been through a lot of editing/rebasing/fixing-typos already.

What I do not know is... when I rebase and then force push... do these comments here still stay (accessible) in the Github interface, given that changed commits have different id's etc.

ThomasWaldmann replied:
not sure whether they stay, just try?

wnienhaus replied:
Ok. Just pushed the rebased code with all fixes so far. And I see these comments remain in this pull-request and the commits they relate to get an "Outdated" tag next to them. So now I know how that works :)

wnienhaus commented:
For now, I have pushed the fixed as additional commits (because I dont know what force pushing will do to this PR's history). But I want to rebase/squash them later, so they "original" commits are clean, without those commits.

ThomasWaldmann commented on tests/assemble.py:
```diff
@@ -121,6 +121,20 @@ def test_assemble_global():
+    a = Assembler()
+    try:
+        a.assemble("  WAIT 42")
+    except Exception as e:
```
can we use a more specific exception here?

wnienhaus replied:
I kept what was raised before, but I guess a `ValueError` exception could be nice here. I will change it.

ThomasWaldmann replied:
ok, great.

ThomasWaldmann commented on tests/assemble.py:
```diff
@@ -121,6 +121,20 @@ def test_assemble_global():
+        if str(e) != "Unknown opcode or directive: WAIT":
+            # re-raise failures we didn't expect
+            raise
+        raised = True
+    else:
+        raised = False
+    assert not raised
```
see `pytest.raises`.

wnienhaus replied:
Do you suggest to add a dependency on `pytest` or to extract the logic into a helper method that is similar to `pytest.raises`? (The latter might be quite nice because I did end up checking for exceptions more than a few times)

Anyway, I followed the pattern that was used in the existing tests, because that worked and used the tooling available.

I'm also not aware of any `pytest` module for MicroPython, and the `unittest` lib that lives in `micropython-lib` is a bit broken. So I was happy with your simple/pragmatic approach to testing for raised exception and duplicated the approach.

Do you feel we should use the opportunity here to improve the testing tooling/framworks used?

ThomasWaldmann replied:
oops, no, we better don't add pytest.

i'm used to it, but maybe better not for micropython.

ThomasWaldmann commented on esp32_ulp/opcodes.py:
```diff
@@ -267,6 +268,22 @@ def make_ins(layout):
+    parts = []
+    for token in split_tokens(arg):
+        if symbols.has_sym(token):
+            _, _, sym_addr = symbols.get_sym(token)
```
sym_value is maybe more suitable as name here.

wnienhaus replied:
Will update.

ThomasWaldmann commented on esp32_ulp/opcodes.py:
```diff
@@ -267,6 +268,22 @@ def make_ins(layout):
+            parts.append(token)
+    parts = "".join(parts)
+    try:
+        parts = str(eval(parts))
```
**Uhoh, guess this is "assembler injection" (see "sql injection").**

ThomasWaldmann replied:
Also, this means that the expression syntax is the same as for python.

Is that compatible to the usual assemblers.

wnienhaus replied:
I did a fair bit of reading up here originally, because I tend to be one of the first to call `eval() == evil()`. In the end I decided that this might be a worthwhile trade-off, because for the expressions I could see being evaluated here the syntax is close enough, and in terms of misuse I found the risk quite low, as the code that gets assembled would be quite short, and python code would be very obvious (and believing code off the net should never be trusted without review anyway).

Do you feel that this is a reasonable assumption, or would you prefer we go a safer route? I will say that I would prefer a different solution, however the only one I could think of for now was to implement the expression parser / evaluator, which would add a lot of code and it may use more memory, thus reducing the size of things we can assemble on-device, and since MicroPython had a way to evaluate things, I used that.

Anyway, about "expression syntax": I had found this: https://pypi.org/project/pcpp/#known-bugs-ordered-from-worst-to-least-worst - a similar project to create a preprocessor. They had concluded that using Python to evaluate expressions worked surprisingly well and I considered their listed edge cases acceptable (perhaps this should be documented).

I see a few ways forward:
1. Do nothing. Accept the risk/imperfection
2. Add a description/warning/listing-of-edge-cases to the documentation
3. Add validation - i.e. tokenise the string-to-be-evaluated and check that it consists only of numbers, brackets, supported arithmetic/logical operators or known symbols.
4. Implement a proper parser/evaluator that does away entirely with calling eval().

Which would you prefer we pick or do you see other options? I am starting to like option 3, which I had not thought of before now.

wnienhaus replied:
I just implemented option 3, as it felt easy enough to do. I have pushed it "on-top" for now (commit dd3a04b), but I intend to squash the change later, once I know you are happy with this option. (hmm, just noticed I don't have a test for the changed "eval_arg". I'll add it)

ThomasWaldmann replied:
sounds reasonable.

ThomasWaldmann commented on esp32_ulp/opcodes.py:
```diff
@@ -267,6 +268,22 @@ def make_ins(layout):
+    parts = "".join(parts)
+    try:
+        parts = str(eval(parts))
+    except:
+        pass
```
aka "the diaper pattern" :-)

wnienhaus replied:
Continuing the pattern that was already there :). I always try to stick with styles/approaches used, rather than to impose my approach, because a file with many different styles mixed is worse in my opinion.

What I do like (a pattern, which I thought I was continuing), is the early-return-fall-through approach, where the last line of the function is the "if all else failed, do this". In case of the try-except-else you proposed further down, there is no last line on it's own but it's the else of a bigger structure. That potentially makes it more likely that one would have to untangle the try-except-else when making future changes (e.g. adding one more case) and require more care to make sure there remains an "if all else failed, do this" case (of course tests should catch this). For that reason, I would choose to keep it as is, but happy to go either way.

ThomasWaldmann replied:
can be done, if it can't be avoided.

about the try/except/else: the point was not only to move the operation into the else, but also to reduce the scope of the try-block to what really should be covered.

wnienhaus replied:
Oh, sorry. I had not noticed that these comments related to two different parts of the code altogether. Somehow Github doesn't show enough context, so it looks like opening the file from the heading of the snippet being shown here, gives a better view. I will use that from now on.

You are very right. Catching anything and throwing it away is not the way to go here. I initially wrapped it that way, because inside eval anything can happen, so I didnt know what exactly to catch, but while adding the validation code yesterday, I actually noticed the try/except around the eval isnt even needed. When we get to that eval, it must evaluate to something or the expression is invalid. So in the latest commit (which I will rebase+squash) I removed this altogether.

ThomasWaldmann commented on esp32_ulp/opcodes.py:
```diff
@@ -289,8 +306,12 @@ def arg_qualify(arg):
         pass
-    entry = symbols.get_sym(arg)
-    return ARG(SYM, entry, arg)
+    try:
+        entry = symbols.get_sym(arg)
+        return ARG(SYM, entry, arg)
+    except KeyError:
+        pass
+    return ARG(IMM, int(eval_arg(arg)), arg)
```
```
    try:
        entry = symbols.get_sym(arg)
    except KeyError:
        return ARG(IMM, int(eval_arg(arg)), arg)
    else:
        return ARG(SYM, entry, arg)
```

wnienhaus replied:
Fixed now.

ThomasWaldmann reviewed (commented):
Reviewed: 
support ULP opcodes in upper case  … 33b4d3d  
add a compatibilty test for the recent fixes and improvements  … 9c94247  
add support for evaluating expressions  … 4962139  
add a compatibilty test for evaluating expressions  … be27cdc  
docs: add that expressions are suppported 521e924

**guess that eval() is dangerous.**

ThomasWaldmann commented on esp32_ulp/util.py:
```diff
@@ -52,3 +52,12 @@ def file_exists(filename):
+def validate_expression(param):
+    for token in split_tokens(param):
+        for c in token:
+            if '0' <= c <= '9' or c in ' \t+-*/%()<>&|^!~xabcdef':
```
```
if c not in ' \t+-*/%()<>&|^!~x0123456789abcdef':
    return False
```

ThomasWaldmann commented on tests/util.py:
```diff
@@ -31,6 +31,27 @@ def test_split_tokens():
+    assert validate_expression('_') is False
+    assert validate_expression('=') is False
+    assert validate_expression('.') is False
+
```
assert validate_expression('evil()') is False

ThomasWaldmann replied:
assert validate_expression('dead()') is False

wnienhaus replied:
Good catch! This was a bit harder to test for than the simple check if a character is in a set of allowed ones, but I now check that characters a-f can only exist in a sequence of valid hex digits that started with 0x. See commit bc6b848

wnienhaus replied:
Btw, I also added tests for `eval_arg` to test that all expressions (operators) we support actually get evaluated correctly. Additionally I have added to the compat tests (expr.S) all supported cases, to ensure that binutils gets to the same result.

(Surprisingly the ^ operator does not result in the same output. My testing so far shows python to get to the right result, while binutils gets to either 0 or 1 depending on what I put in. With all reading so far, I could not figure out why: it should really be treated like a bitwise XOR by binutils, but for some reason isn't. I also grepped the relevant include files from ESP-IDF and they never use ^ in defining constants, so this operator is not so important. Consequently I have removed support for that operator for now. Using it will for now throw the "Unsupported operator" exception)

ThomasWaldmann commented on esp32_ulp/__main__.py:
```diff
@@ -2,11 +2,17 @@
+    preprocessor = Preprocessor()
+    return preprocessor.preprocess(src)
+
+
```
move this to preprocess module?

wnienhaus replied:
I originally had this inline further down as `src = Preprocessor().preprocess(src)` but then extracted it into a method, similar to how src_to_binary is it's own method, and knowing we'd later want to "configure" the Preprocessor instance (see commit d770b98).

It feels a bit weird to move the method, which creates instances of Preprocessor into the preprocess module (i.e. the outside world could call that method many times, creating new instances of Preprocessor all the time, without reusing the instance (that is, unless instantiating Preprocessor outside the module anyway). The singleton pattern could be an option, that the module itself will reuse an existing instance, but since the Preprocessor is a stateful instance, this also feels weird).

What was your main reason for proposing the move?

ThomasWaldmann replied:
well, it is a convenience wrapper and fits there better than in main module.
a lot of python stdlib modules offer some classes but also such convenience wrappers for the usual simple use cases.

wnienhaus replied:
I added a convenience wrapper now. Looks nicer. Also in the later commit when the DefinesDB is added, it was simple enough to just add an `use_defines_db` argument to the convenience function (defaults to `True`).

ThomasWaldmann commented on esp32_ulp/preprocess.py:
```diff
@@ -0,0 +1,53 @@
+from . import nocomment
```
you added a new call to `nocomment`, but I did not see you removing the original call that is already in use somewhere.

wnienhaus replied:
I wanted the pre-processor to be optional - i.e. one can still run the assembler without preprocessing. That's why both the preprocessor and the assembler run nocomment.

Perhaps I can add a `remove_comments = True` arg to the Assembler.assemble() method, i.e. by default comment are removed, but when using the preprocessor before the assembler, one could specify `False` to that arg to tell the assembler to no repeat the comment removal? The input text to that method would still need to be split into lines, so I am not sure how much memory advantage we get, and so far I considered the time dimension (how long the algorithm runs) not that important.

Any suggestion how to proceed?

ThomasWaldmann replied:
not sure about how important time is, depends on how slow we are getting.
it usually is a 1 time op at program startup, so not important to be super fast, but also should not be super slow of course.

but working vs. not working (out of memory) is way more important, so be careful with memory usage. this needs practically trying on esp32, won't find this on dev pc.

ThomasWaldmann replied:
the new arg sounds like a good idea, True is correct default.

wnienhaus replied:
Naming things .... :) 
```diff
 from . import opcodes
-from .nocomment import remove_comments
+from .nocomment import remove_comments as do_remove_comments
 from .util import garbage_collect

 TEXT, DATA, BSS = 'text', 'data', 'bss'
@@ -271,8 +271,8 @@ class Assembler:
                 raise ValueError('Unknown opcode or directive: %s' % opcode)
         self.finalize_sections()

-    def assemble(self, text):
-        lines = remove_comments(text)
+    def assemble(self, text, remove_comments=True):
+        lines = do_remove_comments(text) if remove_comments else text.splitlines()
         self.init(1)  # pass 1 is only to get the symbol table right
         self.assembler_pass(lines)
         self.symbols.set_bases(self.compute_bases())
```
Any suggestion for what else to call the new arg (given that `remove_comments` already is the method we want to call) or what to use instead of the alias? Or happy with the above suggestion?

ThomasWaldmann commented on esp32_ulp/preprocess.py:
```diff
@@ -0,0 +1,53 @@
+
+    def preprocess(self, content):
+        self.parse_defines(content)
+        lines = [l for l in nocomment.remove_comments(content) if l != ""]
```
when implementing the original code, i frequently and quickly ran into memory issues, so IIRC i called remove_comments very early to get rid of most of unneeded string content before doing anything else.

when testing this PR, it should be also tested on a standard esp32 device to see how its runtime behaviour regarding to memory usage / garbage collection is.

especially stuff working on the whole content string or splitting/joining all lines needs quite a bit of memory.
and we don't have gigabytes to use on the esp32...

ThomasWaldmann replied:
btw, if you kill empty lines, line numbers change - do we need them for userfriendly error messages?

wnienhaus replied:
I managed to assemble quite a lot of examples without problem on a standard esp32, including with a database of defines (that is added to the preprocessor in a later commit) containing all ULP/RTC constants from the ESP-IDF.

That all worked for examples up to 6kb in input, thereafter I needed to add an explicit garbage collect - I added it into the `for line in lines` loop below, so that the garbage collector was called after every line. Worked fine but doesnt feel super elegant. I left it out of the pull-request for now, because I wanted to first get feedback on the general design of the solution and then to test/optimise more around memory usage (given that the current implementation managed all but 2 quite big examples I tried).

ThomasWaldmann replied:
are line numbers in error messages still correct if you remove empty lines?

wnienhaus replied:
None of the errors show line numbers of the assembly source (always only the python call stacks), so this is not important. However, I guess empty lines don't take much to hold in memory, so we could keep them too. Maybe better if we ever do add error output related to the input code.

ThomasWaldmann replied:
yeah, i'ld expect the python interpreter to only use 1 empty string object and just refer to that from everywhere.

ThomasWaldmann replied:
oh, i thought we have line numbers in error messages.
didn't use this code for quite a while, sorry.

wnienhaus replied:
I now preserve empty lines. Instead of "not using much more memory", it even resulted in 128 bytes **less** memory being used :) (independent of the size of the input text).

ThomasWaldmann commented on esp32_ulp/preprocess.py:
```diff
@@ -0,0 +1,53 @@
+
+    def parse_defines(self, content):
+        result = {}
+        for line in content.splitlines():
```
avoid by doing nocomment first?

wnienhaus replied:
Good idea, but nocomment would strip out the #define lines, so we need to parse them first.

One could potentially adapt nocomment to accept a list of lines as input, rather than a single string, so that there isnt this back and forth conversion all the time. I had actually thought of that as a future potential improvement to use even less memory - to make all parts capable of operating line-by-line, either using the generator pattern or by returning state (e.g. currently inside a multiline comment) that gets passed into the next call as the starting state, etc..

But for now I have managed to assemble (including the preprocessing) up to 6k of input code on the actual ESP32 which was more than almost all the example code I found and tested with. When doing "aggressive" garbage collect - in the preprocess main loop after every line, then even the biggest file I had worked.

Do you feel this should be improved or ok to leave as is?

ThomasWaldmann replied:
maybe file a ticket about refactoring this later, this PR is already way to big, so we avoid such refactorings here. splitting once and avoiding join/split/join/split.

wnienhaus replied:
Yes. Lets do that. Looks like issue #38 already cover this. Once the preprocess is merged into master, I will update that issue, and add that there's opportunity in the preprocessor for saving memory too.

ThomasWaldmann commented on esp32_ulp/preprocess.py:
```diff
@@ -0,0 +1,53 @@
+        result = {}
+        for line in content.splitlines():
+            line = line.strip()
+            if line[:7] != "#define":
```
line.startswith(...)

wnienhaus replied:
done

ThomasWaldmann commented on esp32_ulp/assemble.py:
```diff
@@ -277,7 +277,10 @@ def assembler_pass(self, lines):
                     func = getattr(opcodes, 'i_' + opcode.lower(), None)
                     if func is not None:
-                        instruction = func(*args)
+                        if self.a_pass == 1:
+                            instruction = 0  # dummy
+                        else:
+                            instruction = func(*args)
```
```
instruction = 0 if self.a_pass == 1 else func(*args)
```

wnienhaus replied:
done

ThomasWaldmann reviewed (commented):
reviewed:
add preprocessor that can replace simple #define values in code  … d5c274f  
fix evaluation of expressions during first assembler pass  … 5480e1d  
remove no-longer-needed pass dependent code from SymbolTable  … 5811060

wnienhaus commented:
For now I am done fixing what I could. Happy to get more feedback.

ThomasWaldmann commented on esp32_ulp/preprocess.py:
```diff
@@ -2,6 +2,46 @@
+        rtc_reg, low_bit, value = args
+        return RTC_Macros.WRITE_RTC_REG((rtc_reg, low_bit, 1, value + ' & 1'))
+
+    @staticmethod
+    def eval_expr(expr):
+        try:
+            return eval(expr)
+        except:
+            return expr
```
unused? if so, please remove.

wnienhaus replied:
Sneaky method: hiding in plain sight! I will remove.
(Originally I did have expression evaluation in the preprocessor - but felt it was better in the assembler, made the preprocessor simpler and by moving it there it allowed for that feature to exist already before the preprocessor).

ThomasWaldmann commented on esp32_ulp/preprocess.py:
```diff
@@ -42,12 +82,33 @@ def expand_defines(self, line):
 
+    def expand_rtc_macros(self, line):
+        clean = line.strip()
+        if len(clean) == 0:
```
if not clean_line:

ThomasWaldmann commented on esp32_ulp/preprocess.py:
```diff
@@ -42,12 +82,33 @@ def expand_defines(self, line):
+            return line
+
+        macro = clean.split('(', 1)
+        if len(macro) != 2:
+            return line
```
add below:
macro_name, macro_args = macro

and then use these instead of macro[0] or splitting again.

wnienhaus replied:
Very nice! Just did that and looks much cleaner. But it doesn't remove a splittng, because the macro_args need to be split further (one reason why i originally didnt like to use an "args" naming for the 2nd part of the split - because initially it's really the "uncleaned args", but it still looks nicer now. Will push soon)

ThomasWaldmann commented on esp32_ulp/preprocess.py:
```diff
@@ -42,12 +82,33 @@ def expand_defines(self, line):
+        if len(macro) != 2:
+            return line
+
+        expand_fn = getattr(RTC_Macros, macro[0], None)
+        if expand_fn is None:
+            return line
```
expand_fn -> macro_fn maybe?

ThomasWaldmann commented on esp32_ulp/preprocess.py:
```diff
@@ -42,12 +82,33 @@ def expand_defines(self, line):
+        _, args = macro
+        args, _ = args.rsplit(')', 1)  # trim away right bracket. safe because comments should already be stripped
+        args = args.split(',')  # not safe when string args contain ',' but we should not have those
+        args = [x.strip() for x in args]
+
+        return expand_fn(args)
```
instead of giving a args list to the fn, you could give *args - and then you could save that args-splitting line in these function bodies by just having separate args instead of the list in the fn header.

wnienhaus replied:
Very nice! Just did that and looks much cleaner.

ThomasWaldmann commented on tests/02_compat_rtc_tests.sh:
```diff
@@ -66,7 +66,7 @@ for src_file in ulptool/src/ulp_examples/*/*.s binutils-esp32ulp/gas/testsuite/g
 
     # for now, skip files that contain known bugs in esp32_ulp (essentially a todo list of what to fix)
-    for I in rtcio esp32ulp_all esp32ulp_globals esp32ulp_jumpr esp32ulp_ranges test_reg; do
+    for I in esp32ulp_all esp32ulp_globals esp32ulp_jumpr esp32ulp_ranges test_reg; do
```
why did you remove rtcio here?

ThomasWaldmann commented on esp32_ulp/opcodes.py:
```diff
@@ -354,7 +354,9 @@ def get_cond(arg):
 def _soc_reg_to_ulp_periph_sel(reg):
     # Map SoC peripheral register to periph_sel field of RD_REG and WR_REG instructions.
     ret = 3
-    if reg < DR_REG_RTCCNTL_BASE:
+    if reg < 0x3ff:
+        ret = 0
```
can you check whether there is a constant we can use instead of 0 and 3?

wnienhaus replied:
Will do. The ret = 3 is actually has no use (ret is always assigned something else), so I have dropped it. Also found a constant name for 0x3ff in the binutils-esp32ulp sources (DR_REG_MAX_DIRECT), so switched to using that too.

ThomasWaldmann commented on esp32_ulp/preprocess.py:
```diff
@@ -78,21 +79,27 @@ def expand_defines(self, line):
+                result = self.parse_define_line(line)
+                db.update(result)
+
+        self.close_db()
```
can we use a context manager for open/close db?

ThomasWaldmann commented on esp32_ulp/util.py:
```diff
@@ -67,3 +68,12 @@ def validate_expression(param):
+
+
+def file_exists(filename):
+    try:
+        os.stat(filename)
+        return True
+    except:
+        pass
+    return False
```
try/except/else? also: the exception raised - is it FileNotFoundError?

wnienhaus replied:
Fixed. Catching OSError now.

ThomasWaldmann commented on tests/preprocess.py:
```diff
@@ -214,6 +219,94 @@ def test_process_include_file_with_multiple_files():
 
 
+@test
+def test_process_include_file_using_database():
```
did you try running these on the esp32 device?

wnienhaus replied:
I have not run the unit tests on the device. I have run all .S files from the compat tests and from the tests I added as compat_rtc_tests on the device though and they worked as expected. (Actually.. I did not then download the resulting binary to compare with binutils - I guess under the assumption that MicroPython on or off the device would run the same.)

What is the reason you are asking? The know that the unit tests can run on the device? Or to ensure it all works in the limited memory (which my testing have shown so far), or something else? Depending on what you were after, I can do more testing.

wnienhaus replied:
Btw, just ran all unit test on the esp32 and they all pass.

ThomasWaldmann replied:
Yay, great.

I was asking because running them on esp32 might show more issues compared to running it on dev PC.

ThomasWaldmann commented:
BTW, i first thought "OMG, this DefinesDB is a bit too much". But considering the memory issues lurking and that the db has to be built only once, guess this is the only non-cumbersome method we can use.

The only (cumbersome) alternative would be to extract "used symbols only" from the large header files.

wnienhaus commented:
> BTW, i first thought "OMG, this DefinesDB is a bit too much". But considering the memory issues lurking and that the db has to be built only once, guess this is the only non-cumbersome method we can use.
> 
> The only (cumbersome) alternative would be to extract "used symbols only" from the large header files.

Actually the btree db will pretty much use constant memory, which can be tuned to stay quite small (for now I let the library pick its own pagesize and cachesize (http://docs.micropython.org/en/latest/library/btree.html?highlight=btree#btree.open), which worked ok.

I also considered: what if we limit everything to "used symbols only" but this brings the challenge, that other things wont assemble. And in case one only needs a handful of constants, one could always #define them in the .S file. To allow for any more flexibility would require all (most) of the symbols somewhere, or the original include files on the device.

I then realised, since the database doesn't use more or less memory depending on how many keys are in it, I could instead focus on keeping the DB as small as possible (use as little filesystem space as possible). I noticed that pretty much all rtc examples include only 3 files ("soc/rtc_cntl_reg.h", "soc/rtc_io_reg.h", "soc/soc_ulp.h") and that indexing them into the db would be enough. Indexing all headers from the esp-idf would result in a 2MB database file, while indexing only those 3 common files results in a 200k database. 200k still feels massive, but I would say that gives the DB it's reason-for-existence, because that would not fit in memory.

(Other things I tried before implementing the DB was to convert the #define lines from include files to python code (something like the current soc.py), but the sheer amount of constants made pretty much anything that puts them in memory impossible. So using the filesystem came as an idea, and I found the btree db fit pretty much perfectly).

So now with the db, which removes memory constraints, with a suitably populated database (i'm thinking build-once-and-forget) the assembler can assemble pretty much any ulp/rtc code.

ThomasWaldmann commented:
OK, looks like I'm through with 1st pass of review. I suggest you keep scope as it is now, but do `git rebase -i`as you desire to collapse some stuff.

When ready, tell me so I can do i 2nd review pass.

wnienhaus commented:
Ok, I have cleaned up everything, and rebased/squashed it all into as clean commits as I could.

I updated the preprocessor docs too, since there's now the convenience function for preprocessing which defaults to using the DefinesDB.

In one of the unit test files, I built an `assert_raises` method, which makes the test read nicer. I did not copy the approach to other unit test files - this could come in a refactoring commit at some later stage.

I also built a context manager for opening and closing the DefinesDB when preprocessing or parsing include files into the defines db. That definitely makes the code look cleaner. Thank you for that suggestion.

You can now take a 2nd pass over the commits - they should be pretty close to what they were like before, but with the improvements embedded into them.

ThomasWaldmann commented:
BTW, please add yourself either to the LICENSE file or add a new file AUTHORS and reference that from LICENSE (see borgbackup/borg for an example).

ThomasWaldmann commented on tests/compat/expr.S:
```diff
@@ -0,0 +1,44 @@
+  move r3, result
+  st r0, r3, 0
+
+  #test that expressions evaluate correctly for all supported operators
+  move r3, 1+2
+  move r3, 3-5
+  move r3, -5
+  move r3, 2*3
+  move r3, 4/2
+  move r3, 4 % 3
+  move r3, 0xff << 2
+  move r3, 0xff >> 1
+  move r3, (0xabcdef | 0xff) & 0xff
+  move r3, 0x1234 & ~2
+  move r3, 42|4&0xf  # 46 (4&0xf is evaluated first)
+  move r3, (42|4)&0xf  # 14 (42|4 is evaluated first)
```
this is a bit strange mix of a real-world code example and some lines of assembler testing.

maybe separate these lines at least?

wnienhaus replied:
Did you mean into a separate file, or maybe I can move them below the `exit: halt` line and add a comment that they are all "artificial" examples?

ThomasWaldmann replied:
whatever you like better.

ThomasWaldmann commented on esp32_ulp/__main__.py:
```diff
@@ -10,7 +10,7 @@
 
 def src_to_binary(src):
     assembler = Assembler()
-    assembler.assemble(src)
+    assembler.assemble(src, remove_comments=False)  # comments already removed by preprocessor
     garbage_collect('before symbols export')
     addrs_syms = assembler.symbols.export()
     for addr, sym in addrs_syms:
```
a typical use of this code is to import src_to_binary from here and call it with some source as str argument.

the main function is not used in this case, so the preprocess step should be moved from there to here.

ThomasWaldmann replied:
the comment then also makes more sense for this use case. :-)

wnienhaus replied:
Changed.

ThomasWaldmann commented on esp32_ulp/assemble.py:
```diff
@@ -277,7 +277,7 @@ def assembler_pass(self, lines):
                     func = getattr(opcodes, 'i_' + opcode.lower(), None)
                     if func is not None:
-                        instruction = func(*args)
+                        instruction = 0 if self.a_pass == 1 else func(*args)
```
as this is not easy to understand, maybe copy some text from commit comment into a source comment above this line.

wnienhaus replied:
Added a description comment

ThomasWaldmann commented on esp32_ulp/preprocess.py:
```diff
@@ -2,6 +2,34 @@
+
+    @staticmethod
+    def WRITE_RTC_REG(rtc_reg, low_bit, bit_width, value):
+        args = (
+            rtc_reg,
+            '%s + %s - 1' % (low_bit, bit_width),
+            low_bit,
+            value
+        )
+        return '\treg_wr ' + ', '.join(args)
```
maybe do it without `args` (see READ_... above).

wnienhaus replied:
:) the READ_.. case above I tried out the different approach, but didn't mean to commit it. But i have changed them both that approach now.

ThomasWaldmann commented on esp32_ulp/preprocess.py:
```diff
@@ -32,29 +32,31 @@ def WRITE_RTC_FIELD(rtc_reg, low_bit, value):
+            # skip defines without value
+            return {}
+        identifier, value = parts
+        tmp = identifier.split('(', 1)
+        if len(tmp) == 2:
```
can it be > 2?

wnienhaus replied:
No, because of the `,1` argument to the split function, which means only split once. This results in either a single value (if there was no way to split) or 2 values (if there was), but never more.

ThomasWaldmann replied:
ah, right, sorry, overlooked that.

ThomasWaldmann commented on tests/fixtures/incl.h:
```diff
@@ -0,0 +1,5 @@
+#define MACRO(x,y) x+y
+#define MULTI_LINE abc \
+                   xyz
+#define CONST2 99
```
there is no newline at the end of the file - diff unhappy.

wnienhaus replied:
Fixed.

ThomasWaldmann commented on tests/fixtures/incl2.h:
```diff
@@ -0,0 +1,2 @@
+#define CONST2 123
+#define CONST3 777
```
there is no newline at the end of the file - diff unhappy.

wnienhaus replied:
Fixed.

ThomasWaldmann commented on esp32_ulp/preprocess.py:
```diff
@@ -103,17 +106,43 @@ def expand_rtc_macros(self, line):
+        self._defines_db = defines_db
+
+    def open_db(self):
+        class ctx:
+            def __init__(self, db):
+                self._db = db
+
+            def __enter__(self):
+                # not opening DefinesDB - it opens itself when needed
+                return self._db
+
+            def __exit__(self, type, value, traceback):
+                if isinstance(self._db, DefinesDB):
+                    self._db.close()
+
+        if self._defines_db:
+            return ctx(self._defines_db)
+
+        return ctx(self._defines)
```
did you try having the contextmanager inside DefinesDB class?

ThomasWaldmann replied:
if it works, also call the context manager from the tests.

wnienhaus replied:
I have tried to put the context manager into the DefinesDB class, but because it handles both the "with DefinesDB" and the "without a DefinesDB" case, this doesn't fit into the DefinesDB class.

I could find an elegant way to add a condition to the "with" statement - ie. only use the context manager from DefinesDB when we're actually using the DefinesDB, otherwise work with the class-internal dict.

I'll mull over this a bit more and see if I can find a way.

Regarding testing: in effect this was tested - because the existing tests exercise the code which uses the context manager, but I added another test now to check specifically that the db is actually closed too.

ThomasWaldmann reviewed (commented):
ok, did a 2nd pass and found some minor stuff.

please only fix these without extending scope, do additional commits for now.

wnienhaus commented on esp32_ulp/assemble.py:
```diff
@@ -93,11 +96,15 @@ def resolve_relative(self, symbol):
 
-    def __init__(self, symbols=None, bases=None):
-        self.symbols = SymbolTable(symbols or {}, bases or {})
+    def __init__(self, symbols=None, bases=None, globls=None):
+        self.symbols = SymbolTable(symbols or {}, bases or {}, globls or {})
```
Hmm. There were two such cases in the file, and I fixed only the first. Well spotted. Fixed.

wnienhaus commented on esp32_ulp/assemble.py:
```diff
@@ -240,7 +240,7 @@ def d_align(self, align=4, fill=None):
 
     def d_set(self, symbol, expr):
-        value = int(expr)  # TODO: support more than just integers
+        value = int(opcodes.eval_arg(expr))  # TODO: support more than just integers
```
Removed.

wnienhaus commented on esp32_ulp/util.py:
```diff
@@ -2,10 +2,68 @@
+    tokens = []
+    state = NORMAL
+    for c in line:
+        if ('a' <= c <= 'z') or ('A' <= c <= 'Z') or ('0' <= c <= '9') or c == '_':
```
Changed it. Oi - that was a lot of typing :) . and afterwards I noticed i missed the W letter :) . Somehow I feel "safer" with the original approach, but I have no idea which executes "better".

wnienhaus reviewed (commented):
Made all the fixes now (except the Context Manger move to the DefinesDB class, which I need to think over more). I'll push this as a commit on top now, and then check the Context Manager topic.

wnienhaus commented:
Pushed latest improvements as a commit on top: 5fd45a2

ThomasWaldmann reviewed (approved):
LGTM - merge and do the remaining stuff in later PRs?

wnienhaus commented:
Ok. I force pushed one small thing - moving the "made up" statement in tests/compat/expr.S to the bottom of the file with an improved comment.

And I updated the LICENCE file and added an AUTHORS file. Perhaps check if you are happy with that, and then we're good for a merge.

(For now I don't manage to find a solution to the Context Manager challenge, but that can be improved later)

ThomasWaldmann commented on AUTHORS:
```diff
@@ -0,0 +1,23 @@
+source, and for expanding RTC convenience macros in a way that mirrors what the real
+ESP-IDF macros do.
+Copyright 2021, Wilko Nienhaus <wilko.nienhaus@gmail.com>, licenced under the terms
+of The MIT License (MIT).
```
diff unhappy, missing newline at end of file.

wnienhaus replied:
Hehe :) will fix.

ThomasWaldmann commented on LICENSE:
```diff
@@ -1,6 +1,7 @@
 The MIT License (MIT)
 
-Copyright (c) 2018 Thomas Waldmann
+Copyright (c) 2018 Thomas Waldmann (see AUTHORS file)
+Copyright (c) 2021 Wilko Nienhaus (see AUTHORS file)
```
Maybe it is simpler to update authors only at one place, so I suggest this:

Copyright 2018-2021 by the py-esp32-ulp authors, see AUTHORS file.

Also, I would not go into details about who has written what, that's where git log is better anyway and a file describing this will likely be always outdated.

wnienhaus replied:
Done. 47d5e8a

ThomasWaldmann commented:
Yay! \o/

wnienhaus commented:
Yay! And thank you for all your effort with this big PR!
